crates/matrix-sdk-base/src/response_processors/state_events.rs

```diff
@@ -76,22 +86,71 @@ pub mod sync {
         room_info: &mut RoomInfo,
         ambiguity_cache: &mut AmbiguityCache,
         new_users: &mut U,
+        state_store: &BaseStateStore,
     ) -> StoreResult<()>
     where
         U: NewUsers,
     {
         for (raw_event, event) in iter::zip(raw_events, events) {
-            room_info.handle_state_event(event);
-
-            if let AnySyncStateEvent::RoomMember(member) = event {
-                dispatch_room_member(
-                    context,
-                    &room_info.room_id,
-                    member,
-                    ambiguity_cache,
-                    new_users,
-                )
-                .await?;
+            match event {
+                AnySyncStateEvent::RoomMember(member) => {
+                    room_info.handle_state_event(event);
+
+                    dispatch_room_member(
+                        context,
+                        &room_info.room_id,
+                        member,
+                        ambiguity_cache,
+                        new_users,
+                    )
+                    .await?;
+                }
+
+                AnySyncStateEvent::RoomCreate(create) => {
+                    if super::is_create_event_valid(
+                        context,
+                        room_info.room_id(),
+                        create,
+                        state_store,
+                    ) {
+                        room_info.handle_state_event(event);
+                    } else {
+                        error!(
+                            room_id = ?room_info.room_id(),
+                            ?create,
+                            "`m.create.tombstone` event is invalid, it creates a loop"
+                        );
+
+                        // Do not add the event to `room_info`.
+                        // Do not add the event to `context.state_changes.state`.
+                        break;
```

Comment:
Hmm, if we break here, won't this potentially ignore a bunch of valid events as well?

Was that supposed to be a continue?

Reply:
This function checks for a single new state event, so at worst, it skips a single state event.